articles/external_qos_file.md

```diff
@@ -0,0 +1,402 @@
+
+A `rcl_qos_loaded_profiles_t` member could be added in both `rcl_init_options_t` and `rcl_node_options_t`.
+The one in init options will be common to all nodes in the same contexts, and can be overridden by the one in node options.
+
+Optionally, we could also add in `rcl_node_options_t` a boolean to indicate to use the `rcl_qos_loaded_profiles_t` in init options if the profile was not found in the node specific loaded profiles object.
+
+### New ros argument
+
+A new argument will be needed to allow the user to specify which file to load.
+
+```bash
+ros2run <package_name> <exec_name> --ros-args --qos-file /path/to/qos/file
+```
+
+This loading mechanism will be equivalent to setting the `rcl_qos_loaded_profiles_t` member in init options.
+
+## Further extensions
+
+### Loading qos profiles in composable nodes
+
+The interface definition of the [LoadNode](https://github.com/ros2/rcl_interfaces/blob/master/composition_interfaces/srv/LoadNode.srv) service could be extended with a field to specify the desired `QoS` of the loaded node.
+
+In that way, the user loading the node is responsible of passing the desired QoS, instead of the node container.
+The message to be sent could be loaded from a qos profile file.
+
+#### Proposed interfaces
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
+```
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
+```
+
+```
+# composition_interfaces/srv/LoadNode.srv
+
+# Current request fields
+...
+# New field
+EndpointQoS[] requested_qos_profiles  # new field to allow loading user requested qos profiles
+---
+# Current response definition
+...
+```
+
+A `rcl_qos_loaded_profiles_t` can be constructed from the received message and passed to the node options of the node to be loaded.
+
+TBD: `QoS` message definition.
+
+### Support in launch files
+
+`launch_ros.actions.Node` and `launch_ros.actions.LoadComposableNode` actions could support a qos file argument in their constructors.
+
+## File format
+
+As showed in the above examples, both YAML or XML can correctly represent the desired file format.
+
+XML advantages:
+- Can be easily verified.
+YAML advantages:
+- It can be argued that the format is simpler and more readable than XML.
+- The format proposed in the above examples can easily be integrated with the current parameter file format.
+  That integration would lead in a unique "configuration file", that could be further extended in the future.
+  For example, passing remapping rules in it.
+
+## Explicitly allowing external configurability in entities
+
+Both node and init options could have a flag to ignore the new `--qos-file` argument.
```

Comment:
I'm not convinced we want to completely ignore QoS overrides. If a user tries to set the QoS for a node that doesn't accept it, I don't think it should silently fail. They should see at least a warning, or at most an error should occur when starting the node. This avoids a situation where someone thinks that they've changed the QoS when it didn't actually do anything.

I realize this gets a little tricky if we want to allow "default" QoS settings for all nodes (as described earlier in the document). Perhaps in this case, it is okay to silently ignore QoS settings not tied to named node. On the other hand, it could still be misleading and maybe it's still better to warn or error. I'm ambivalent at the moment.

Reply:
I come down on the side of "if the system integrator says it, the node can't ignore it".

Nodes need a way to validate that the provided QoS meets the requirements of the node. e.g. A SLAM node needs to get every laser scan so lossy QoS won't do. But apart from that nodes should either accept the given QoS or they should noisily reject it.

Reply:
> I'm not convinced we want to completely ignore QoS overrides. If a user tries to set the QoS for a node that doesn't accept it, I don't think it should silently fail. They should see at least a warning, or at most an error should occur when starting the node. This avoids a situation where someone thinks that they've changed the QoS when it didn't actually do anything.

I don't think it should fail, I think it should not try to get the QoS settings from a file.

> They should see at least a warning

The problem with showing a warning, is that you're actually loading and checking the file.

> Nodes need a way to validate that the provided QoS meets the requirements of the node. e.g. A SLAM node needs to get every laser scan so lossy QoS won't do. But apart from that nodes should either accept the given QoS or they should noisily reject it.

I don't see why the node's author should reject a non reliable QoS in this case.
If the person integrating the system wants to try best effort communication, they should be able to do so.

Reply:
> If the person integrating the system wants to try best effort communication, they should be able to do so.

This sounds contrary to the idea that node authors have to opt-in. If an integrator wants to use best effort for a topic, and the author hasn't opted-in to allow this change to be made, then I think the integrator should receive a warning.

> The problem with showing a warning, is that you're actually loading and checking the file.

Maybe it doesn't make sense to have an option to ignore the QoS file. From an integrator's perspective, if I provide a QoS file trying to configure a node, I want to know if that configuration is valid (ie. if the node actually accepted the changes to QoS). So, I think it's necessary for something to parse the file and check with the relevant nodes that everything is okay, warning otherwise.

Reply:
> I don't see why the node's author should reject a non reliable QoS in this case.
If the person integrating the system wants to try best effort communication, they should be able to do so.

I think it's like saying "You can use character buffers without null if you like, but the behaviour is undefined" versus "You can't use unterminated strings". We can give node authors the ability to specify the limits on their node's capabilities and _have the tool chain enforce them_, or we can rely on developers to follow what the documentation says and hope they realise it's their use of the node that's broken, not the node's implementation, when it doesn't work.

This is particularly important for a node author who wants to provide guarantees about behaviour for reasons such as safety.

> Maybe it doesn't make sense to have an option to ignore the QoS file. From an integrator's perspective, if I provide a QoS file trying to configure a node, I want to know if that configuration is valid (ie. if the node actually accepted the changes to QoS). So, I think it's necessary for something to parse the file and check with the relevant nodes that everything is okay, warning otherwise.

I agree with this. I don't like giving nodes the ability to blanket-reject all QoS changes. I'd prefer node authors to have to at least think about it, and if they still blanket-reject all changes they have to write that themselves and we can see it as a bad smell.

Reply:
We can add a callback to reject the changes if that's what the majority wants.

IMO, it's adding a "false sense" of ensuring that the QoS will be used in the correct way, and we can end up with node's authors restricting QoS settings that actually work unnecessarily (e.g. restricting only to the ones they tested).
The mechanism to ensure that the proposed QoS is valid is IMO extremely verbose too.

With "false sense", I mean that the system integrator can still screw up the QoS between nodes.
They could setup a reliable QoS in a subscription in one node, and configure by mistake the "matching" publisher with best effort reliability, and they system won't work.

For example in this case:
>  A SLAM node needs to get every laser scan so lossy QoS won't do

Should the node author reject all QoS that are non reliable and doesn't have keep all durability?
Because without both, the communication can't be ensured to be truly non-lossy.

> Maybe it doesn't make sense to have an option to ignore the QoS file. From an integrator's perspective, if I provide a QoS file trying to configure a node, I want to know if that configuration is valid (ie. if the node actually accepted the changes to QoS). So, I think it's necessary for something to parse the file and check with the relevant nodes that everything is okay, warning otherwise.

:+1:, always parsing the file sounds reasonable .

> I don't like giving nodes the ability to blanket-reject all QoS changes

I'm sorry, but I don't understand where "blanket-reject all QoS changes" was suggested.
Maybe I miss-explained something.

Reply:
> I'm sorry, but I don't understand where "blanket-reject all QoS changes" was suggested.

Maybe this is in response to my comment about requiring node authors to opt-in to QoS configuration? I do think that the default behavior should be to "blanket-reject" all QoS changes. I.e. If the node author doesn't explicitly configure any node entities to have configurable QoS, then any attempts to override QoS values should fail (IMO, not silently).

Reply:
> With "false sense", I mean that the system integrator can still screw up the QoS between nodes.
They could setup a reliable QoS in a subscription in one node, and configure by mistake the "matching" publisher with best effort reliability, and they system won't work.

This is true, and it hints at the two different levels to what is allowable QoS:

1. Nodes need to define what they can support to ensure their behaviour is as defined by the node author. In general, I think we prefer that node authors go as broad as possible in what QoS they support.
2. System integrators need to define what they want to use. In general, they should go as narrow as possible.

I do agree that there is a risk that a node author will just shut it down to only what they've tested, but I don't see how that is different from a default of "reject all changes". However in either case the most important thing to me is that it fails noisily, not silently.

Reply:
I've added the concept of QoS acceptance callbacks to the proposal https://github.com/ros2/design/pull/296/commits/a0e635f3500e457bf56a36862651fe3bcf50b0a7.

> However in either case the most important thing to me is that it fails noisily, not silently.

The idea is that if the loaded QoS profile is rejected, the code will fail immediately.